### PR TITLE
proposal: resource distribution

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: veophi <vec.g.sun@gmail.com>

### Ⅰ. Describe what this PR does
proposal: resource distribution

### CODE REVIEW

File: docs/proposals/20210713-resourcedistribution.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+#### Story 1: distribute to other namespaces
+```yaml
+# my-secret.yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: namespaced-secret
+  annotations:
+    openkruise.io/sync-to: "namespaces"
+    openkruise.io/sync-to/namespaces: "ns-1;ns-2;ns-3"
```
Comment (zmberg) [inline docs/proposals/20210713-resourcedistribution.md:61]: 1. openkruise.io/sync-to/namespaces -> e.g. openkruise.io/sync-to-namespaces
2. whether the user can configure only one annotations, e.g. openkruise.io/sync-to-namespaces
3. support all-namespaces, except kube-system

File: docs/proposals/20210713-resourcedistribution.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  password: t0p-Secret
+```
+When `kubectl apply -f my-secret.yaml`,  the secret will be created or updated in **all namespaces**.
+
+### Implementation
+We will watch all the events about the `Secret`, and sync it when reconciling it.
+The main logic looks like this:
+
+#### WebHook Validation
+The value　of `Annotations["openkruise.io/sync-to"]` is limited in `["cluster", "namespace"]` , and it should be validated.
```
Comment (zmberg) [inline docs/proposals/20210713-resourcedistribution.md:88]: Is there no need to do webhook validate checks in this place

**Review by FillZpp — COMMENTED** [review]
Is it a good idea to implement resource distribution through the annotations or labels?

If so, we should specify the kinds that support to distribute, and kruise-manager has to list watch all objects of these kinds.

Comparing to define a CRD named ResourceDistribution and users put the object into CR, which design will be better?

**Comment by hantmac** [comment]
> Is it a good idea to implement resource distribution through the annotations or labels?
> 
> If so, we should specify the kinds that support to distribute, and kruise-manager has to list watch all objects of these kinds.
> 
> Comparing to define a CRD named ResourceDistribution and users put the object into CR, which design will be better?

In my opinion, using annotations or labels is more convenient and easy to use,  and the new CRD requires learning costs for users. But define a CRD is easier to implement and has more clear code.

File: docs/proposals/20210713-resourcedistribution.md (hunk 2 above)
Comment (veophi) [reply docs/proposals/20210713-resourcedistribution.md:88]: ok

**Review by furykerry — APPROVED** [review]

**Pull request merged**